lib/AST/DiagnosticConsumer.cpp

```diff
@@ -14,15 +14,148 @@
+  }
+#endif
+}
+
+void
+FileSpecificDiagnosticConsumer::buildLocationToConsumerMap(SourceManager &SM) {
+  using MapEntry = std::pair<CharSourceRange, unsigned>;
+
+  // Look up each file's source range and add it to the "map" (to be sorted).
+  size_t i = 0;
+  for (const ConsumerPair &pair : SubConsumers) {
+    SWIFT_DEFER { ++i; };
+    if (pair.first.empty())
+      continue;
+
+    Optional<unsigned> bufferID = SM.getIDForBufferIdentifier(pair.first);
+    assert(bufferID.hasValue() && "consumer registered for unknown file");
+    CharSourceRange range = SM.getRangeForBuffer(bufferID.getValue());
+    LocationToConsumerMap.emplace_back(range, i);
+  }
+
+  // Sort the "map" by buffer /end/ location, for use with std::lower_bound
+  // later. (Sorting by start location would produce the same sort, since the
+  // ranges must not be overlapping, but since we need to check end locations
+  // later it's consistent to sort by that here.)
+  std::sort(LocationToConsumerMap.begin(), LocationToConsumerMap.end(),
+            [](const MapEntry &left, const MapEntry &right) -> bool {
+    auto compare = std::less<const char *>();
+    return compare(getRawLoc(left.first.getEnd()).getPointer(),
+                   getRawLoc(right.first.getEnd()).getPointer());
+  });
+
+  // Check that the ranges are non-overlapping.
+  assert(LocationToConsumerMap.end() ==
+           std::adjacent_find(LocationToConsumerMap.begin(),
+                              LocationToConsumerMap.end(),
+                              [](const MapEntry &left, const MapEntry &right) {
+                                return left.first.overlaps(right.first);
+                              }) &&
+         "overlapping ranges despite having distinct files");
+}
+
+DiagnosticConsumer *
+FileSpecificDiagnosticConsumer::consumerForLocation(SourceManager &SM,
+                                                    SourceLoc loc) const {
+  if (loc.isInvalid())
+    return nullptr;
+
+  // This map is generated on first use and cached, to allow the
+  // FileSpecificDiagnosticConsumer to be set up before the source files are
+  // actually loaded.
+  if (LocationToConsumerMap.empty()) {
+    auto *mutableThis = const_cast<FileSpecificDiagnosticConsumer*>(this);
+    mutableThis->buildLocationToConsumerMap(SM);
+  }
+
+  // "Find the first range whose end location is greater than or equal to
+  // 'loc'."
+  auto possiblyContainingRangeIter =
+      std::lower_bound(LocationToConsumerMap.begin(),
+                       LocationToConsumerMap.end(),
+                       loc,
+                       [](const std::pair<CharSourceRange, unsigned> &entry,
+                          SourceLoc loc) -> bool {
+    auto compare = std::less<const char *>();
+    return compare(getRawLoc(entry.first.getEnd()).getPointer(),
+                   getRawLoc(loc).getPointer());
+  });
+
+  if (possiblyContainingRangeIter != LocationToConsumerMap.end() &&
```

Comment:
My style would be to factor out this predicate into a function called `actuallyContains` or some such.

Reply:
Since it takes all the context of the current method (both the LocationToConsumerMap and the location we're searching for), I don't think that pulls its weight. Doubly so because iterators are weird annoying types to spell out.

Reply:
I think the new comment at the top of the lower_bound covers this now.